capsules/src/process_console.rs

```diff
@@ -120,6 +144,64 @@ pub struct ProcessConsole<'a, A: Alarm<'a>, C: ProcessManagementCapability> {
     capability: C,
 }
 
+#[derive(Copy, Clone)]
+pub struct Command {
+    buf: [u8; COMMAND_BUF_LEN],
+    len: usize,
+}
+
+impl Command {
+    /// Write the buffer with the provided data.
+    /// If the provided data's length is smaller than the buffer length,
+    /// the left over bytes are not modified due to '\0' termination.
+    fn write(&mut self, buf: &[u8; COMMAND_BUF_LEN]) {
+        self.len = buf
+            .iter()
+            .position(|a| *a == EOL)
+            .unwrap_or(COMMAND_BUF_LEN);
+
+        (&mut self.buf).copy_from_slice(buf);
+    }
+
+    fn insert_byte(&mut self, byte: u8) {
+        if self.len < COMMAND_BUF_LEN {
+            self.buf[self.len] = byte;
```

Comment:
Not blocking, but in these type of cases I feel it's better to use `get()/get_mut()`.

So for example (untested):

```rust
if Ok(b) = self.get_mut(self.len) {
    *b = byte;
}
```

That avoids the panic. I would assume that rustc is smart enough here to understand that this can't panic and optimise it out, but AFAIK there is no guarantee that it will

Reply:
I do not know if rustc will optimize this, becasue the `insert` and `delete` methods are called when user is typing a character or is pressing `backspace` in order to delete a character, so if the command buffer is empty (delete case) and user is pressing a backspace this would cause a panic and the same way for the insert case.

However I changed the code for the `insert_byte` and `delete_last_byte` to match the suggestion.